libs/utils/src/filter.c

```diff
@@ -17,421 +17,497 @@
+static char* celix_filter_parseAttributeOrValue(const char* filterString, int* pos, bool parseAttribute) {
+    const char* name = parseAttribute ? "attribute" : "attribute value";
+    celix_autofree char* value = NULL;
+    size_t valueSize = 0;
+    celix_autoptr(FILE) stream = NULL;
 
+    bool keepRunning = true;
     while (keepRunning) {
         char c = filterString[*pos];
 
         switch (c) {
-            case ')': {
-                keepRunning = 0;
-                break;
-            }
-            case '(': {
-                fprintf(stderr, "Filter Error: Invalid value.\n");
-                free(value);
+        case ')': {
+            if (parseAttribute) {
+                celix_err_pushf("Filter Error: Unexpected `)` char while parsing %s.", name);
                 return NULL;
             }
-            case '\0':{
-                fprintf(stderr, "Filter Error: Unclosed bracket.\n");
-                free(value);
-                return NULL;
+            keepRunning = false;
+            break;
+        }
+        case '(': {
+            celix_err_pushf("Filter Error: Unexpected `(` char while parsing %s.", name);
+            return NULL;
+        }
+        case '\0': {
+            celix_err_pushf("Filter Error: Unexpected end of string while parsing %s.", name);
+            return NULL;
+        }
+        case '<':
+        case '>':
+        case '=':
+        case '~': {
+            if (parseAttribute) {
+                keepRunning = false; // done for attribute, valid for value
+                break;
             }
-            case '\\': {
-                (*pos)++;
+            /* no break */
+        }
+        default: {
+            if (c == '\\') {
+                if (filterString[*pos + 1] == '\0') {
+                    celix_err_pushf("Filter Error: Unexpected end of string while parsing %s.", name);
+                    return NULL;
+                }
+                (*pos)++; // eat '\'
                 c = filterString[*pos];
             }
-            /* no break */
-            default: {
-                char ch[2];
-                ch[0] = c;
-                ch[1] = '\0';
-                strcat(value, ch);
-                (*pos)++;
-                break;
+            if (!stream) {
+                stream = open_memstream(&value, &valueSize);
+                if (!stream) {
+                    celix_err_push("Filter Error: Failed to open mem stream.");
+                    return NULL;
+                }
+            }
+            int rc = fputc(c, stream);
+            if (rc == EOF) {
+                celix_err_push("Filter Error: Failed to write to stream.");
+                return NULL;
             }
+            (*pos)++;
+            break;
+        }
         }
     }
 
-    if (strlen(value) == 0) {
-        fprintf(stderr, "Filter Error: Missing value.\n");
-        free(value);
+    if (!stream) {
+        celix_err_push("Filter Error: Empty value.\n");
```

Comment:
According to [java.util.Properties.load(java.io.Reader)](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/util/Properties.html#load(java.io.Reader)) 

>  As a third example, the line:
cheeses
specifies that the key is "cheeses" and the associated element is the empty string "". 

If empty value is not allowed, we cannot match key with empty value using equal filters.

Reply:
updated to allow empty attribute value